GVFS/GVFS.Platform.Windows/WindowsFileSystemVirtualizer.cs

```diff
@@ -1248,7 +1268,7 @@ private HResult NotifyPreDeleteHandler(string virtualPath, bool isDirectory)
                     }
                     else
                     {
-                        this.OnWorkingDirectoryFileOrFolderDeleteNotification(virtualPath, isDirectory, isPreDelete: false);
+                        this.OnWorkingDirectoryFileOrFolderDeleteNotification(virtualPath, isDirectory);

```

Comment:
I like that this change keeps the call to `OnPossibleTombstoneFolderCreated` in `WindowsFileSystemVirtualizer`, but I'm not sure about all of the back and forth that we know have between FileSystemCallbacks and the Mac/WindowsFileSystemVirtualizer.cs:

- `ProjFS` -> `WindowsFileSystemVirtualizer` -> `FileSystemCallbacks` -> `WindowsFileSystemVirtualizer` -> `FileSystemCallbacks`

What if this call to `OnWorkingDirectoryFileOrFolderDeleteNotification` was replaced with something like the following (which could always go in a new `WindowsFileSystemVirtualizer` if that would be cleaner):

```
            if (isDirectory)
            {
                GitCommandLineParser gitCommand = new GitCommandLineParser(this.Context.Repository.GVFSLock.GetLockedGitCommand());
                if (!gitCommand.IsValidGitCommand)
                {
                    this.FileSystemCallbacks.OnFolderDeleted(relativePath);
                }
                else
                {
                    // During a git command if it deletes a folder we need to track that as a tombstone
                    // So that we can delete them if the projection changes
                    this.FileSystemCallbacks.OnPossibleTombstoneFolderCreated(relativePath);
                }
            }
            else
            {
                this.FileSystemCallbacks.OnFileDeleted(relativePath);
            }

            this.FileSystemCallbacks.InvalidateGitStatusCache();
```

And we removed `OnWorkingDirectoryFileOrFolderDeleteNotification` and the new `OnFileDeleteNotification` and `OnFolderDeleteNotification` methods?

Similarly in `MacFileSystemVirtualizer.cs` the call to `OnWorkingDirectoryFileOrFolderDeleteNotification` could be replaced with:

```
            if (isDirectory)
            {
                GitCommandLineParser gitCommand = new GitCommandLineParser(this.Context.Repository.GVFSLock.GetLockedGitCommand());
                if (!gitCommand.IsValidGitCommand)
                {
                    this.FileSystemCallbacks.OnFolderPreDelete(relativePath);
                }
            }
            else
            {
                this.FileSystemCallbacks.OnFilePreDelete(relativePath);
            }

            this.FileSystemCallbacks.InvalidateGitStatusCache();
```

If we wanted to share a bit of code we could always add a common helper method like this:

```
bool IsGitCommandRunning()
{
    GitCommandLineParser gitCommand = 
        new GitCommandLineParser(this.Context.Repository.GVFSLock.GetLockedGitCommand());
    return gitCommand.IsValidGitCommand;
}
```

This does duplicate the decision of whether OnFolderDelete should be called (based on if the git command is running), but it avoids having to add the new `OnFileDeleteNotification` and `OnFolderDeleteNotification` methods.

I think given the choice between the current changes and the other alternative you suggested (no-op `OnPossibleTombstoneFolderCreated` in Mac) my preference is the latter, as I think it might be tougher to understand why Mac and Windows call different methods in their `OnFileDeleteNotification` and `OnFolderDeleteNotification` overrides.

Reply:
I originally did refactor it like you are suggesting but the issue I saw was that the were both making the same decisions and it was only the last call to `FileSystemCallbacks` that was platform specific.  So if we needed to add something like the invalidate of the git status cache we would need to remember to add it in both places which I didn't like because sooner or later when you have to remember, you will forget until there is an issue with a user.  

I will go with the no-op then because it really is code that I think should be shared if at all possible which we can to avoid bugs because we forgot to change something in multiple places.

Reply:
> I will go with the no-op then because it really is code that I think should be shared if at all possible which we can to avoid bugs because we forgot to change something in multiple places.

Ok thanks that sounds good.